src/main/drivers/pwm_output_dshot.c

```diff
@@ -96,6 +108,115 @@ void pwmWriteDshotInt(uint8_t index, uint16_t value)
     }
 }
 
+#ifdef USE_DSHOT_TELEMETRY
+uint32_t readDoneCount;
+
+static void motor_DMA_IRQReadHandler(dmaChannelDescriptor_t *descriptor)
+{
+    if (DMA_GET_FLAG_STATUS(descriptor, DMA_IT_TCIF)) {
+        motorDmaOutput_t * const motor = &dmaMotors[descriptor->userParam];
+
+        motor->hasTelemetry = true;
+        DMA_Cmd(motor->timerHardware->dmaRef, DISABLE);
+        TIM_DMACmd(motor->timerHardware->tim, motor->timerDmaSource, DISABLE);
+        DMA_CLEAR_FLAG(descriptor, DMA_IT_TCIF);
+
+        readDoneCount++;
+    }
+}
+
+static uint16_t decodeDshotPacket(uint32_t buffer[])
+{
+    uint32_t value = 0;
+    for (int i = 1; i < DSHOT_TELEMETRY_INPUT_LEN; i += 2) {
+        int diff = buffer[i] - buffer[i-1];
+        value <<= 1;
+        if (diff > 0) {
+            if (diff >= 11) value |= 1;
+        } else {
+            if (diff >= -9) value |= 1;
+        }
+    }
+    
+    uint32_t csum = value;
+    csum = csum ^ (csum >> 8); // xor bytes
+    csum = csum ^ (csum >> 4); // xor nibbles
+
+    if (csum & 0xf)
+        return 0xffff;
+    return value >> 4;
+}
+
+static uint16_t decodeProshotPacket(uint32_t buffer[])
+{
+    uint32_t value = 0;
+    for (int i = 1; i < PROSHOT_TELEMETRY_INPUT_LEN; i += 2) {
+        const int proshotModulo = MOTOR_NIBBLE_LENGTH_PROSHOT;
+        int diff = ((buffer[i] + proshotModulo - buffer[i-1]) % proshotModulo) - PROSHOT_BASE_SYMBOL;
+        int nible;
+        if (diff < 0) {
+            nible = 0;
+        } else {
+            nible = (diff + PROSHOT_BIT_WIDTH / 2) / PROSHOT_BIT_WIDTH;
+        }
+        value <<= 4;
+        value |= nible;
+    }
+
+    uint32_t csum = value;
+    csum = csum ^ (csum >> 8); // xor bytes
+    csum = csum ^ (csum >> 4); // xor nibbles
+
+    if (csum & 0xf)
+        return 0xffff;
+    return value >> 4;
+}
+
+
+void pwmStartDshotMotorUpdate(uint8_t motorCount)
+{
+    if (dshotTelemetryEnabled) {
+        for (int i = 0; i < motorCount; i++) {
+            if (dmaMotors[i].hasTelemetry) {
+                uint16_t value = dmaMotors[i].useProshot ?
+                    decodeProshotPacket(
+                        dmaMotors[i].dmaBuffer ) :
+                    decodeDshotPacket(
+                        dmaMotors[i].dmaBuffer );
+                if (value != 0xffff) {
+                    dmaMotors[i].dshotTelemetryValue = value;
+                    if (i < 4) {
+                        DEBUG_SET(DEBUG_RPM_TELEMETRY, i, value);
+                    }
+                } else {
+                    dshotInvalidPacketCount++;
+                    if (i == 0) {
+                        memcpy(inputBuffer,dmaMotors[i].dmaBuffer,sizeof(inputBuffer));
```

Comment:
This is only for debug purposes and only doing it for one motor prevents too many cycles to be lost here. This will likely get taken out once everything has become stable.